test/format/format.sh

```diff
@@ -0,0 +1,301 @@
+		grep 'caught signal' $log > /dev/null && {
+			rm -rf $dir $log
+			[[ $verbose -ne 0 ]] && echo "$0: job in $dir aborted"
+			continue
+		}
+
+		# Report failures.
+		grep 'RUN PARAMETERS' $log > /dev/null && {
+			echo "$0: failure status reported" > $dir/reported
+			failure=$(($failure + 1))
+			echo "$0: job in $dir failed"
+		}
+	done
+	return 0
+}
+
+# Start a single job.
+count_runs=0
+run_format()
+{
+	count_runs=$(($count_runs + 1))
+	dir="$home/RUNDIR.$count_runs"
+	log="$dir.log"
+
+	if [[ $smoke_test -ne 0 ]]; then
+		args=${smoke_list[$smoke_next]}
+		smoke_next=$(($smoke_next + 1))
+		echo "$0: starting smoke-test job in $dir"
+	else
+		args=$format_args
+
+		# If abort/recovery testing is configured, do it 5% of the time.
+		[[ $abort_test -ne 0 ]] && [[ $(($count_runs % 20)) -eq 0 ]] && args="$args abort=1"
+
+		echo "$0: starting job in $dir"
+	fi
+	[[ $verbose -ne 0 ]] &&
+	    echo "$0: running $format_binary -c "$config" -h "$dir" -1 $args quiet=1"
+	$format_binary -c "$config" -h "$dir" -1 $args quiet=1 > $log 2>&1 &
+}
+
+seconds=$((minutes * 60))
+start_time="$(date -u +%s)"
+while :; do
+	# Check if our time has expired.
+	[[ $seconds -ne 0 ]] && {
+		now="$(date -u +%s)"
+		elapsed=$(($now - $start_time))
+
+		# If we've run out of time, terminate all running jobs.
+		[[ $elapsed -ge $seconds ]] && {
+			stop=1
+			forcequit=1
+		}
+	}
+
+	# Start more jobs.
+	while :; do
+		# Check if we're only running the smoke-tests and we're done.
+		[[ $smoke_test -ne 0 ]] && [[ $smoke_next -ge ${#smoke_list[@]} ]] && stop=1
+	
+		# Check if the total number of jobs to run has been reached.
+		[[ $runs -ne 0 ]] && [[ $count_runs -ge $runs ]] && stop=1
+
+		# Check if less than 60 seconds left on any timer. The goal is to avoid killing
+		# jobs that haven't yet configured signal handlers, because we rely on handler
+		# output to determine their final status.
+		[[ $seconds -ne 0 ]] && [[ $(($seconds - $elapsed)) -lt 60 ]] && stop=1
+
+		[[ $stop -ne 0 ]] && break;
+
+		# Check if the maximum number of jobs in parallel has been reached.
+		n=`jobs -p | wc -l`
+		[[ $n -ge $jobs ]] && break
+
+		# Start another job, but don't pound on the system.
+		run_format
+		sleep 2
+	done
+
+	# Forcibly quit in some cases.
+	[[ $forcequit -ne 0 ]] && {
+		pkill --signal TERM -P $$
+		sleep 5
+	}
+
+	# Wait for any completed jobs.
+	children=$(pgrep -P $$)
+	for i in $children; do
+		kill -s 0 $i || wait -n
+	done
+
+	# Clean up and update status.
+	success_save=$success
+	failure_save=$failure
+	resolve
+	[[ $success -ne $success_save ]] || [[ $failure -ne $failure_save ]] &&
+	    echo "$0: $success successful runs, $failure failed runs"
+
+	# Quit if we're done and there aren't any jobs left to wait for.
+	children=$(pgrep -P $$)
+	[[ $stop -ne 0 ]] && [[ -z "$children" ]] && break;
+
+	# Forcibly quit if there's a failure and first-failure configured.
+	[[ $failure -ne 0 ]] && [[ $first_failure -ne 0 ]] && {
+		forcequit=1
+		stop=1
```

Comment:
There are multiple instances of setting both flags in the script. I guess the idea here is to make flag setting and handling unsynchronised?

Reply:
Yes; there's a worker loop that handles killing running processes and waiting for them, let me know if there's a better way to do this.

If we're willing, for example, to guarantee that `GNU parallel` or `xargs` are installed everywhere we want to run this script, we might be able to simplify things.

Reply:
I don't think `GNU parallel` is preinstalled on all Evergreen distros, and it's better not to depend on that.